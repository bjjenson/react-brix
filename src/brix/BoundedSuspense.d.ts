export interface ISuspense {
  fallback: React.ReactType<React.ReactPropTypes>
  boundary: React.ReactType<React.ReactPropTypes>
}

declare const BoundedSuspense: React.ComponentType<ISuspense>
export default BoundedSuspense
