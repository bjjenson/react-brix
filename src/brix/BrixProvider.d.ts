import * as React from 'react'
import { Map } from 'immutable'

export interface BrixProps extends React.HTMLAttributes<HTMLDivElement> {
  value: any,
}

export type BrixProvider = React.ComponentType<BrixProps>
export function getBrixContext(): React.Context<{ state: Map<string, any> }>
export function getBrixConsumer(): React.Consumer<BrixProps>
