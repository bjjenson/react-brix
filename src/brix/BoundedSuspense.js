import React from 'react'
import PropTypes from 'prop-types'

class BoundedSuspense extends React.Component {
  state = { hasError: false }

  static getDerivedStateFromError(error) {
    return { hasError: true }
  }

  componentDidCatch(error, info) {
    console.log(error, info)
  }

  render() {
    const { fallback, boundary, children } = this.props
    const { hasError } = this.state
    if (hasError) {
      return boundary
    }

    return (
      <React.Suspense fallback={fallback}>
        {children}
      </React.Suspense>
    )
  }
}

BoundedSuspense.propTypes = {
  fallback: PropTypes.node.isRequired,
  boundary: PropTypes.node,
}

BoundedSuspense.defaultProps = {
  boundary: <h3 style={{ color: 'red' }}>Something is bad</h3>,
}

export default BoundedSuspense