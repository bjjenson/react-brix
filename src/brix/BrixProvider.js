import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { Map } from 'immutable'

const BrixContext = React.createContext()

export const BrixProvider = ({ value, children }) => {
  const [state, setState] = useState(value)

  return (
    <BrixContext.Provider value={{ state, setState }}>
      {children}
    </BrixContext.Provider>
  )
}

BrixProvider.propTypes = {
  value: PropTypes.object,
}

BrixProvider.defaultProps = {
  value: Map(),
}

export function getBrixContext() {
  return BrixContext
}

export function getBrixConsumer() {
  return BrixContext.Consumer
}
