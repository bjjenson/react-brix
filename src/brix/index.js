export { BrixConsumer, BrixProvider, BrixContext } from './BrixProvider'
export { default as BoundedSuspense } from './BoundedSuspense'
export { useBrix } from './useBrix'
export { useBrixSetter } from './useBrixSetter'
export { useBrixWorker } from './useBrixWorker'
export { useWorker, getKey } from './useWorker'
