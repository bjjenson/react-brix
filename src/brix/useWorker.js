let index = 0
const cache = {}

export const useWorker = (key, worker) => {
  if (cache[key]) {
    const workedItem = cache[key]
    if (workedItem.error) {
      throw workedItem.error
    }

    if (workedItem.result) {
      return workedItem.result
    }
    throw workedItem.worker
  } else {
    const workedItem = {
      worker: worker()
        .then(response => {
          workedItem.result = response
          return response
        }, rejection => {
          console.log('rejected', rejection)
          workedItem.error = rejection
        }).catch(err => {
          workedItem.error = err
          console.log('caught', err)
        }),
    }
    cache[key] = workedItem
    throw workedItem.worker
  }
}

export const getKey = () => {
  return `_key${index++}`
}