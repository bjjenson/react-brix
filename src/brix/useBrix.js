import { useContext } from 'react'
import { getBrixContext } from './BrixProvider'
import { coerceDataToImmutable } from './helpers'

export const useBrix = (path, notSetValue) => {
  const { state, setState } = useContext(getBrixContext())
  const value = state.getIn(path, notSetValue)

  const update = updatedValue => {
    setState(currState => currState.setIn(path, coerceDataToImmutable(updatedValue)))
  }

  return {
    value,
    set: update,
    cnx: {
      value,
      onChange: e => update(e.target.value),
    },
  }
}
