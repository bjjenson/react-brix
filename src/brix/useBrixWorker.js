import { useContext } from 'react'
import { useBrix } from './useBrix'
import { useWorker } from './useWorker'
import { BrixContext } from './BrixProvider'

export const useBrixWorker = (path, worker, notSetValue, transformer = v => v) => {
  const { value, set } = useBrix(path)

  if (value === undefined) {

    const getState = () => {
      const { state } = useContext(BrixContext)
      return state
    }

    const result = useWorker(path, () => worker(getState))
    set(transformer(result))

    return {
      value: notSetValue,
    }
  }

  return { value }
}
