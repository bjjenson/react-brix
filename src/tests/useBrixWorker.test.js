import { Map } from 'immutable'
import { useContext } from 'react'
import { useBrix } from '../brix/useBrix'
import { getBrixContext } from '../brix/BrixProvider'
import { useWorker } from '../brix/useWorker'
import { useBrixWorker } from '../brix/useBrixWorker'

jest.mock('react')
jest.mock('../brix/useBrix')
jest.mock('../brix/useWorker')
jest.mock('../brix/BrixProvider')

const set = jest.fn()
const worker = jest.fn()

let path
beforeEach(() => {
  path = ['path', 'to', 'value']
  useBrix.mockReturnValue(['something set', set])
  useContext.mockReturnValue({ state: 'context-state' })
  getBrixContext.mockReturnValue('brix-context')
})

test('returns value from brix', () => {
  const actual = useBrixWorker(path, worker, Map())

  expect(actual).toMatchSnapshot()
  expect(useBrix).toHaveBeenCalledWith(path)
  expect(useContext).toHaveBeenCalledWith('brix-context')
})

test('brix value is undefined return notSetValue', () => {
  useBrix.mockReturnValue([undefined, set])

  const actual = useBrixWorker(path, worker, Map())
  expect(actual).toMatchSnapshot()
  expect(useWorker.mock.calls[0]).toMatchSnapshot()
})

describe('worker', () => {
  beforeEach(() => {
    useBrix.mockReturnValue([undefined, set])
    useWorker.mockImplementation((path, todo) => {
      todo()
      return 'workerResult'
    })
  })

  test('useWorker calls worker', () => {
    useBrixWorker(path, worker, Map())

    expect(worker.mock.calls[0]).toMatchSnapshot()
    expect(set).toHaveBeenCalledWith('workerResult')
  })

  test('worker result is undefined, set notSetValue', () => {
    useWorker.mockImplementation((path, todo) => {
      todo()
      return undefined
    })

    useBrixWorker(path, worker, Map())

    expect(set).toHaveBeenCalledWith(Map())
  })

  test('set uses transformer', () => {
    useBrixWorker(path, worker, Map(), result => `${result} transformed`)

    expect(set).toHaveBeenCalledWith('workerResult transformed')
  })
})
