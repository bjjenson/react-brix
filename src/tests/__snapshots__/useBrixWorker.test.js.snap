// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`brix value is undefined return notSetValue 1`] = `Immutable.Map {}`;

exports[`brix value is undefined return notSetValue 2`] = `
Array [
  Array [
    "path",
    "to",
    "value",
  ],
  [Function],
]
`;

exports[`returns value from brix 1`] = `"something set"`;

exports[`worker useWorker calls worker 1`] = `
Array [
  [Function],
]
`;
